airflow_ext/gfw/models.py

```diff
@@ -90,5 +92,51 @@ def source_table_sensors(self, dag):
             for parts in self.source_table_parts(date=self.source_sensor_date_nodash())
         ]
 
+    def gcs_sensor(self, dag, bucket, prefix, date):
+        return GoogleCloudStoragePrefixSensor(
+            dag=dag,
+            task_id='source_exists_{}'.format(bucket),
+            bucket=bucket,
+            prefix='{}/{}'.format(prefix, date),
+            mode='reschedule',      # the sensor task frees the worker slot when the criteria is not yet met
+                                    # and it's rescheduled at a later time.
+            poke_interval=10 * 60,  # check every 10 minutes.
+            timeout=60 * 60 * 24    # timeout of 24 hours.
+        )
+
+    def source_path(self, date=None):
+        gcs_paths = self.config.get('source_paths') or self.config.get('source_path')
+        assert gcs_paths
+        paths = gcs_paths.split(',')
+        gcs=None
+        tables=None
+
+        for path in paths:
+            if (path.strip().startswith('gs://')):
+                gcs = yield dict(
+                    bucket=re.search('(?<=gs://)[^/]*', path).group(0),
+                    prefix=re.search('(?<=gs://)[^/]*/(.*)', path).group(1),
+                    date=self.source_date_range()[1] if not date else date
+                )
+            else:
+                tables = yield dict(
+                    project=self.config['project_id'],
+                    dataset=self.config['source_dataset'],
+                    table=path,
+                    date=self.source_sensor_date_nodash()
+                )
+        [gcs, tables]
+
+    def source_sensors(self, dag, date=None):
+        return [
+            self.gcs_sensor(dag=dag, **parts) if ('bucket' in parts and 'prefix' in parts)
```

Comment:
I don't like the idea of mixing the BQ and GCS sensors, I think you are adding code complexity for a case that won't happen any time soon, having both BQ and GCS sources for a given pipeline. I would add the new logic in `source_gcs_sensors`.

Reply:
Understood.

Reply:
Yes, better separate the code in 2 functions. And the Dag that extends DagFactory invoke the method that it needs